helper_crates/vtable/src/vrc.rs

```diff
@@ -0,0 +1,206 @@
+    Copyright (c) 2020 Olivier Goffart <olivier.goffart@sixtyfps.io>
+    Copyright (c) 2020 Simon Hausmann <simon.hausmann@sixtyfps.io>
+
+    SPDX-License-Identifier: GPL-3.0-only
+    This file is also available under commercial licensing terms.
+    Please contact info@sixtyfps.io for more information.
+LICENSE END */
+
+//! implementation of vtable::Vrc
+
+use core::cell::Cell;
+
+use super::*;
+
+/// This trait is implemented by the `#[vtable]` macro.
+///
+/// It is implemented if the macro has a "drop_in_place" function.
+pub unsafe trait VTableMetaDropInPlace: VTableMeta {
+    /// Safety: the Target needs to be pointing to a valid allocated pointer
+    unsafe fn drop_in_place(vtable: &Self::VTable, ptr: *mut u8) -> vrc::Layout;
+    unsafe fn dealloc(vtable: &Self::VTable, ptr: *mut u8, layout: vrc::Layout);
+}
+
+/// Similar to [`std::alloc::Layout`], but `repr(C)`
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct Layout {
+    pub size: usize,
+    pub align: usize,
+}
+
+impl From<core::alloc::Layout> for Layout {
+    fn from(layout: core::alloc::Layout) -> Self {
+        Self { size: layout.size(), align: layout.align() }
+    }
+}
+
+impl core::convert::TryFrom<Layout> for core::alloc::Layout {
+    type Error = core::alloc::LayoutErr;
+
+    fn try_from(value: Layout) -> Result<Self, Self::Error> {
+        Self::from_size_align(value.size, value.align)
+    }
+}
+
+#[repr(C)]
+struct VRcInner<'vt, T: VTableMeta, X> {
+    vtable: &'vt T::VTable,
+    strong_ref: Cell<u32>,
+    weak_ref: Cell<u32>,
+    data: X,
+}
+
+/// A reference counted pointer to an object matching the virtual table `T`
+///
+/// Similar to [`std::rc::Rc`] where T is a VTable struct. But it has the following difference with
+/// the type from std:
+/// - It doesn't implement `Deref`, but you need to call `borrow()` to access a `VRef<R>`
+/// which implements `Deref`.
+/// - It does not allow to access mutable reference. (No `get_mut` or `make_mut`), meaning it is
+/// safe to get a Pin reference with `borrow_pin`.
+/// - It is safe to pass it accross ffi boundaries.
+#[repr(transparent)]
+pub struct VRc<T: VTableMetaDropInPlace + 'static> {
+    inner: NonNull<VRcInner<'static, T, u8>>,
+}
+
+impl<T: VTableMetaDropInPlace + 'static> Drop for VRc<T> {
+    fn drop(&mut self) {
+        unsafe {
+            let inner = self.inner.as_ref();
+            inner.strong_ref.set(inner.strong_ref.get() - 1);
+            if inner.strong_ref.get() == 0 {
+                let data = &inner.data as *const u8 as *mut u8;
+                let layout = T::drop_in_place(inner.vtable, data);
+                inner.weak_ref.set(inner.weak_ref.get() - 1);
+                if inner.weak_ref.get() == 0 {
+                    let vtable = inner.vtable;
+                    T::dealloc(vtable, self.inner.cast().as_ptr(), layout);
+                } else {
+                    self.inner.cast::<VRcInner<T, Layout>>().as_mut().data = layout;
```

Comment:
Just for sanity's sake, is there a way to assert/ensure that the size of T is at least as big as the size of Layout?

Reply:
We allocate enough room in VRc::new.  I added a comment